File: packages/typescript/tests/proto-roundtrip.test.ts
// Why: This file pins the wire-format contract. Every fixture in
// tests/conformance/fixtures must encode to byte-identical bytes across
// languages. JS map insertion order + @bufbuild/protobuf's map encoding
// makes this non-trivial — sorted-key pre-insertion is the fix.
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { dirname } from 'node:path';

import { toProto, fromProto, encodeCapturedError, decodeCapturedError } from '../src/proto/adapter.js';
import { emptyCapturedError, type CapturedError } from '../src/types.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const FIXTURES = join(__dirname, '..', '..', '..', 'tests', 'conformance', 'fixtures');

interface FixtureDescriptor {
  captured_error?: {
    type?: string;
    message?: string;
    frames?: Array<Record<string, unknown>>;
    mechanism?: Record<string, unknown> | null;
    release?: string;
    server_name?: string;
  };
}

function loadFixture(name: string): CapturedError {
  const json = JSON.parse(readFileSync(join(FIXTURES, `${name}.json`), 'utf8')) as FixtureDescriptor;
  const c = json.captured_error ?? {};
  const base = emptyCapturedError();
  return {
    ...base,
    type: c.type ?? '',
    message: c.message ?? '',
    release: c.release ?? '',
    server_name: c.server_name ?? '',
    frames: (c.frames ?? []).map((f) => ({
      function: (f.function as string) ?? '',
      module: (f.module as string) ?? '',
      package: (f.package as string) ?? '',
      file: (f.file as string) ?? '',
      abs_path: (f.abs_path as string) ?? '',
      line: (f.line as number) ?? 0,
      context_line: (f.context_line as string) ?? '',
      pre_context: (f.pre_context as string[]) ?? [],
      post_context: (f.post_context as string[]) ?? [],
      source_link: (f.source_link as string) ?? '',
      in_app: (f.in_app as boolean) ?? false,
    })),
    mechanism: c.mechanism
      ? {
          type: (c.mechanism.type as string) ?? '',
          description: (c.mechanism.description as string) ?? '',
          handled: (c.mechanism.handled as boolean) ?? false,
          synthetic: (c.mechanism.synthetic as boolean) ?? false,
          help_link: (c.mechanism.help_link as string) ?? '',
          source: (c.mechanism.source as string) ?? '',
          exception_id: (c.mechanism.exception_id as number) ?? 0,
          parent_id: (c.mechanism.parent_id as number) ?? 0,
          is_exception_group: (c.mechanism.is_exception_group as boolean) ?? false,
          data: ((c.mechanism.data as Record<string, string>) ?? {}),
        }
      : undefined,
  };
}

const FIXTURE_NAMES = [
  '0001-simple',
  '0002-empty-chain',
  '0003-three-deep',
  '0004-source-context',
  '0005-mechanism-data',
];

describe('proto byte-equivalence', () => {
  for (const name of FIXTURE_NAMES) {
    it(`fixture ${name} encodes to canonical bytes`, () => {
      // Why: cross-language conformance — bytes must match Rust's prost output
      // for every fixture. Map encoding determinism is the load-bearing detail.
      const plain = loadFixture(name);
      const actual = encodeCapturedError(plain);
      const expected = readFileSync(join(FIXTURES, `${name}.pb`));
      expect(Buffer.from(actual)).toEqual(expected);
    });
  }

  it('default CapturedError encodes to empty bytes', () => {
    // Why: encoding.feature "Empty CapturedError encodes as empty bytes".
    const bytes = encodeCapturedError(emptyCapturedError());
    expect(bytes.length).toBe(0);
  });

  it('round-trips a fixture through encode + decode', () => {
    // Why: encoding.feature "Default values round-trip".
    const plain = loadFixture('0001-simple');
    const bytes = encodeCapturedError(plain);
    const decoded = decodeCapturedError(bytes);
    expect(decoded).toEqual(plain);
  });

  it('toProto preserves mechanism.data sort order at the wire', () => {
    // Why: JS objects preserve insertion order; @bufbuild/protobuf encodes
    // maps in iteration order. Pre-sorting keys yields the sorted-key wire
    // bytes Rust's BTreeMap produces. This test asserts the adapter does it.
    const plain: CapturedError = {
      ...emptyCapturedError(),
      mechanism: {
        type: '',
        description: '',
        handled: false,
        synthetic: false,
        help_link: '',
        source: '',
        exception_id: 0,
        parent_id: 0,
        is_exception_group: false,
        data: { z: '1', a: '2', m: '3' },
      },
    };
    const proto = toProto(plain);
    const keys = Object.keys(proto.mechanism!.data);
    expect(keys).toEqual(['a', 'm', 'z']);
  });

  it('fromProto round-trips a plain capture', () => {
    // Why: bidirectional fidelity — fromProto(toProto(x)) === x for valid inputs.
    const plain = loadFixture('0005-mechanism-data');
    const proto = toProto(plain);
    const back = fromProto(proto);
    expect(back).toEqual(plain);
  });
});

File: packages/csharp/tests/ProtoRoundtripTest.cs
// Proto round-trip: plain → proto → bytes → proto → plain.
// Also asserts byte-equivalence against the committed conformance
// fixtures, which is the cross-language wire-format contract.

using System.IO;
using System.Linq;
using FluentAssertions;
using Google.Protobuf;
using Sererr;
using Sererr.Proto;
using Xunit;

namespace Sererr.Tests;

public class ProtoRoundtripTest
{
    [Fact]
    // Plain → proto → plain preserves every field.
    public void RoundTrip_PreservesFields()
    {
        var input = new CapturedError
        {
            Type = "MyError",
            Message = "oops",
            Frames = new List<StackFrame>
            {
                new()
                {
                    Function = "f", Module = "mypkg", File = "src/x.rs",
                    Line = 12, InApp = true,
                },
            },
            Mechanism = new ExceptionMechanism
            {
                Type = "generic", Handled = true,
            },
            Release = "v1.0.0",
            ServerName = "host-1",
        };

        var proto = ProtoAdapter.ToProto(input);
        var bytes = proto.ToByteArray();
        var decoded = Sererr.V1.CapturedError.Parser.ParseFrom(bytes);
        var roundtripped = ProtoAdapter.FromProto(decoded);

        roundtripped.Should().Be(input);
    }

    [Fact]
    // Default CapturedError serializes to empty bytes (all zero-value
    // fields are dropped by proto3).
    public void DefaultCapturedError_SerializesEmpty()
    {
        var c = new CapturedError();
        var proto = ProtoAdapter.ToProto(c);
        proto.ToByteArray().Should().BeEmpty();
    }

    [Fact]
    // Map data with multiple keys must serialize sorted-by-key so the
    // bytes are deterministic across languages.
    public void MechanismData_DeterministicByteOrder()
    {
        var a = new CapturedError
        {
            Mechanism = new ExceptionMechanism
            {
                Type = "signal",
                Data = new SortedDictionary<string, string>
                {
                    ["zulu"] = "1",
                    ["alpha"] = "2",
                    ["mike"] = "3",
                },
            },
        };
        // Re-build identical content via a differently-iteration-ordered
        // dictionary; the resulting proto bytes must be identical.
        var b = new CapturedError
        {
            Mechanism = new ExceptionMechanism
            {
                Type = "signal",
                Data = new SortedDictionary<string, string>
                {
                    ["mike"] = "3",
                    ["alpha"] = "2",
                    ["zulu"] = "1",
                },
            },
        };

        var ba = ProtoAdapter.ToProto(a).ToByteArray();
        var bb = ProtoAdapter.ToProto(b).ToByteArray();
        ba.Should().Equal(bb);
    }

    [Theory]
    [InlineData("0001-simple")]
    [InlineData("0002-empty-chain")]
    [InlineData("0003-three-deep")]
    [InlineData("0004-source-context")]
    [InlineData("0005-mechanism-data")]
    // Byte-equivalence with the committed cross-language .pb fixtures.
    // SERERR_FIXTURES_DIR points at tests/conformance/fixtures.
    public void Fixture_BytesMatch(string name)
    {
        var dir = Environment.GetEnvironmentVariable("SERERR_FIXTURES_DIR")
                  ?? FindRepoFixturesDir();
        var json = File.ReadAllText(Path.Combine(dir, $"{name}.json"));
        var expected = File.ReadAllBytes(Path.Combine(dir, $"{name}.pb"));

        var captured = FixtureLoader.LoadCapturedError(json);
        var bytes = ProtoAdapter.ToProto(captured).ToByteArray();

        bytes.Should().Equal(expected, $"fixture {name} must match");
    }

    private static string FindRepoFixturesDir()
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir is not null && !Directory.Exists(Path.Combine(dir.FullName, "tests", "conformance", "fixtures")))
        {
            dir = dir.Parent;
        }
        return Path.Combine(dir!.FullName, "tests", "conformance", "fixtures");
    }
}
